src/v/resource_mgmt/storage.cc

```diff
@@ -218,6 +218,141 @@ eviction_policy::partition* eviction_policy::schedule::current() {
+     */
+    const auto collection_threshold = [this] {
+        const auto& lm = _storage->local().log_mgr();
+        if (!lm.config().delete_retention().has_value()) {
+            return model::timestamp(0);
+        }
+        const auto now = model::timestamp::now().value();
+        const auto retention = lm.config().delete_retention().value().count();
+        return model::timestamp(now - retention);
+    };
+
+    gc_config cfg(
+      collection_threshold(),
+      _storage->local().log_mgr().config().retention_bytes());
+
+    /*
+     * in smallish batches partitions are queried for their reclaimable
+     * segments. all of this information is bundled up and returned.
+     */
+    fragmented_vector<partition> res;
+    co_await ss::max_concurrent_for_each(
+      partitions.begin(), partitions.end(), 20, [&res, cfg](const auto& p) {
+          auto log = dynamic_cast<storage::disk_log_impl*>(p->log().get_impl());
+          auto gate = log->gate().hold(); // protect against log deletes
+          return log->get_reclaimable_offsets(cfg)
+            .then([&res, p](auto offsets) {
+                res.push_back({
+                  .group = p->group(),
+                  .offsets = std::move(offsets),
+                });
+            })
+            .finally([g = std::move(gate)] {});
+      });
+
+    /*
+     * sorting by raft::group_id would provide a more stable round-robin
+     * evaluation of partitions in later processing of the result, but on small
+     * timescales we don't expect enough churn to make a difference, nor with a
+     * large number of partitions on the system would the non-determinism
+     * in the parallelism above in max_concurrent_for_each be problematic.
+     */
+
+    vlog(rlog.trace, "Reporting reclaim offsets for {} partitions", res.size());
+    co_return res;
+}
+
+ss::future<> eviction_policy::install_schedule(schedule schedule) {
+    /*
+     * install the schedule on each core in parallel
+     */
+    auto total = co_await ss::map_reduce(
+      schedule.shards,
+      [this](auto& shard) { return install_schedule(std::move(shard)); },
+      size_t(0),
+      std::plus<>());
+
+    vlog(rlog.info, "Requested truncation from {} cloud partitions", total);
+}
+
+// the per-shard counterpart of install_schedule(schedule)
+ss::future<size_t> eviction_policy::install_schedule(shard_partitions shard) {
+    const auto shard_id = shard.shard;
+    return _pm->invoke_on(shard_id, [shard = std::move(shard)](auto& pm) {
+        size_t decisions = 0;
+        for (const auto& partition : shard.partitions) {
+            if (!partition.decision.has_value()) {
+                continue;
+            }
+
+            auto p = pm.partition_for(partition.group);
+            if (!p) {
+                continue;
+            }
+
+            auto log = dynamic_cast<storage::disk_log_impl*>(
+              p->log().get_impl());
+            log->set_cloud_gc_offset(partition.decision.value());
```

Comment:
this fires an assert when cloud retention is not active, should there be a conditional?

Reply:
can you elaborate on this? the assertions exist, but we also have checks that don't report any reclaimable segments for topics without could retention enabled. maybe there is tho a race condition if the config is dynamically enabled?

Reply:
ok i see now. i changed `set_cloud_gc_offset` to log an error message and be a noop for non cloud topics. assert was a bit extreme. it shouldn't be called here on non-cloud topics because we don't collect reclaimable segments from the non-cloud partitions (ie won't ever reach this), but... it's all based on dynamic configs. so i wouldn't be surprised if there was some narrow race that occurred when configs were changed at runtime.